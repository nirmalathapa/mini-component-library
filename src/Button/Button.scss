@import "../variables.scss";
@import "../typography.scss";

@mixin button--($variants) {
  background: $variants;
  &:hover {
    transition: all 0.1s ease;
  }
}

.button {
  @include font-defaults;
  background: $primary-500;
  color: $white;
  text-align: center;
  border: 1px solid transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  font-size: 1rem;
  display: inline-flex;
  align-items: center;
  justify-content: center;

  width: max-content;
  padding: 8px 16px;
  border-radius: 4px;
  cursor: default;
  user-select: none;
  text-decoration: none;

  &::-moz-focus-inner {
    border-style: "none"; // Remove Firefox dotted outline.
  }
  &:focus,
  &:active,
  &:focus-visible {
    box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.05), 0px 0px 0px 2px $white,
      0px 0px 0px 4px $primary-500;
    outline: transparent;
  }
  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
  .icon {
    margin-right: 8px;
  }
}

.btn--primary {
  @include button--($primary-500);
  border: 1px solid $primary-500;

  &:hover {
    background: $primary-700;
  }
}

.btn--secondary {
  border: 1px solid transparent;
  @include button--($secondary-100);
  color: $primary-700;
  &:hover {
    background: $secondary-200;
  }
}

.btn--tertiary {
  color: $primary-700;
  background-color: transparent;
  border: 1px solid $primary-500;
  &:hover {
    background: $primary-500;
    color: $white;
    .icon path {
      stroke: $white;
    }
  }
  &:focus,
  &:focus-visible {
    .icon path {
      stroke: $white;
    }
  }
}

.btn--md {
  padding: 12px 24px;
  font-size: 1.2rem;
}

.btn--lg {
  padding: 16px 24px;
  font-size: 1.3rem;
}
