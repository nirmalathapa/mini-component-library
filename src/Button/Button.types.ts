import React from "react";

export type CSSColorVar = "primary" | "secondary" | "tertiary" | "white";
export type ButtonSize = "sm" | "md" | "lg";
export type ButtonVariant = "primary" | "secondary" | "tertiary";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children?: React.ReactNode;
  size?: ButtonSize;
  kind?: CSSColorVar;
  disabled?: boolean;
  iconName?: string;
  iconLeft?: React.ReactNode;
}
