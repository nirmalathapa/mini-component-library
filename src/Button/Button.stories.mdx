<!-- Badge.stories.mdx

import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import Button from "./Button";

<Meta
  title="MDX/Button"
  argTypes={{
    table: {
      defaultValue: {
        summary: "positive",
      },
      type: {
        summary: "Shows options to the Badge",
        detail: "Listing of available options",
      },
    },
  }}
  component={Button}
/>

export const Template = (args) => <Button {...args} />;

# Button

Buttons are used to initialize an action. Button labels express what action will occur when the user interacts with it.
Buttons are clickable elements that are used to trigger actions. They communicate calls to action to the user and allow users to interact with pages in a variety of ways. Button labels express what action will occur when the user interacts with it.

## When to use

Use buttons to communicate actions users can take and to allow users to interact with the page. Each page should have only one primary button, and any remaining calls to action should be represented as lower emphasis buttons.

## When not to use

Do not use buttons as navigational elements. Instead, use links when the desired action is to take the user to a new page.

## Button variants

Each button variant has a particular function and its design signals that function to the user. It is therefore very important that the different variants are implemented consistently across products, so that they message the correct actions.
Button variants

<Canvas>
  <Story
    name="Primary"
    args={{
      size: "small",
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Secondary"
    args={{
      size: "small",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
-->
