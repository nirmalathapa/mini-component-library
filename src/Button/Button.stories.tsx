// Button.stories.ts|tsx

import React from "react";
import { action } from "@storybook/addon-actions";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import {
  ArgsTable,
  Stories,
  Heading,
  Title,
  Subtitle,
} from "@storybook/addon-docs/blocks";
import Button from "./Button";
import { Icon } from "../Icons/Icons";

const onClick = action("button-click");

export default {
  title: "Button",
  component: Button,
  parameters: {
    source: true,
    controls: { hideNoControlsWarning: true },
    actions: {
      handles: ["click"],
    },
    docs: {
      page: () => (
        <>
          <Title>Submit Button</Title>
          <Subtitle>
            A button element that is used for intiating the change of the
            context. Icon is used with the button label to give an additional
            clarity that is relevant to the action. The button component is also
            an extension of html button element. which means it can be used as a
            normal html button.
          </Subtitle>
          <Heading>Component API</Heading>
          <ArgsTable />
          <Stories title="Examples" includePrimary />
        </>
      ),
    },
  },

  argTypes: {
    children: {
      description:
        "Specify text label for the button that is of ReactNode type",
      control: false,
    },
    kind: {
      description: "Specify the kind of Button you want to create",
      control: {
        type: "select",
        options: ["primary", "secondary", "tertiary"],
      },
    },
    size: {
      description: `Specify the kind of Button you want to create`,
      control: {
        type: "select",
        options: ["sm", "md", "lg"],
      },
    },
    disabled: {
      description: `Specify the kind of Button you want to create`,
      control: {
        type: "boolean",
        defaultValue: false,
      },
    },
    iconLeft: {
      description: `Provide svg icon component to the left of the button`,
      control: false,
    },
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => (
  <Button {...args} children={<span>Submit</span>} />
);
export const Primary = Template.bind({});
Primary.args = {
  kind: "primary",
  iconLeft: <Icon name="FormIcon" size="sm" colorVar="primary" />,
};

Primary.parameters = {
  docs: {
    storyDescription:
      "Primary button gives the user a visual cue that the action they are taking is important and should be taken seriously. Since primary buttons are for important action, it is recommended not to use multiple times in one screen.",
  },
};

export const Secondary = Template.bind({});
Secondary.args = {
  kind: "secondary",
  iconLeft: <Icon name="FormIcon" size="sm" colorVar="secondary" />,
};

Secondary.parameters = {
  docs: {
    storyDescription:
      "Secondary button are used for low emphasis or not mandatory actions. It is pared with primary buttons to provide secondary action.",
  },
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  kind: "tertiary",
  iconLeft: <Icon name="FormIcon" size="sm" colorVar="tertiary" />,
};

Tertiary.parameters = {
  docs: {
    storyDescription:
      "Teritary buttons are used for less prominent and an independant actions that can be used across different places in the web application. ",
  },
};

export const SetOfButtons = () => {
  return (
    <>
      <Button
        kind="secondary"
        iconLeft={<Icon name="FormIcon" size="sm" colorVar="secondary" />}
      >
        Submit
      </Button>
      <span style={{ padding: "0 5px" }} />
      <Button
        kind="primary"
        iconLeft={<Icon name="FormIcon" size="sm" colorVar="primary" />}
        onClick={onClick}
      >
        Submit
      </Button>
    </>
  );
};
SetOfButtons.parameters = {
  docs: {
    storyDescription:
      "You can use two Button elements where the user has more than one option for a given action. Use the secondary button for less important actions, and the primary button for more important actions.",
  },
};

export const DisabledButtons = () => {
  return (
    <>
      <Button
        kind="secondary"
        disabled={true}
        iconLeft={<Icon name="FormIcon" size="sm" colorVar="secondary" />}
      >
        Submit
      </Button>
      <span style={{ padding: "0 5px" }} />
      <Button
        kind="tertiary"
        disabled={true}
        iconLeft={<Icon name="FormIcon" size="sm" colorVar="tertiary" />}
        onClick={onClick}
      >
        Submit
      </Button>
      <span style={{ padding: "0 5px" }} />

      <Button
        kind="primary"
        disabled={true}
        iconLeft={<Icon name="FormIcon" size="sm" colorVar="primary" />}
        onClick={onClick}
      >
        Submit
      </Button>
    </>
  );
};

DisabledButtons.parameters = {
  docs: {
    storyDescription:
      "A disabled button is used when the user is not allowed to perform the action due to the the context of the application. For example, a disabled button is used when the user is not allowed to perform the action because the user is not logged in.",
  },
};

export const ButtonSize = () => {
  return (
    <>
      <Button
        kind="primary"
        iconLeft={<Icon name="FormIcon" size="sm" colorVar="primary" />}
      >
        Submit
      </Button>
      <span style={{ padding: "0 5px" }} />
      <Button
        kind="primary"
        size="md"
        iconLeft={<Icon name="FormIcon" size="md" colorVar="primary" />}
      >
        Submit
      </Button>
      <span style={{ padding: "0 5px" }} />
      <Button
        kind="primary"
        size="lg"
        iconLeft={<Icon name="FormIcon" size="md" colorVar="primary" />}
      >
        Submit
      </Button>
    </>
  );
};

ButtonSize.parameters = {
  docs: {
    storyDescription:
      "This prop specifies which size the Button should be rendered at. Valid values are field, md, lg. If no size is specified, it renders as default.",
  },
};
