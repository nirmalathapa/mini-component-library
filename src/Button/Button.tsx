import React from "react";
import cx from "classnames";
import { ButtonProps } from "./Button.types";

import "./Button.scss";

const Button: React.FC<ButtonProps> = ({
  size = "sm",
  kind = "primary",
  disabled = false,
  children,
  iconLeft,
}) => (
  <button
    type="button"
    data-testid="button"
    disabled={disabled}
    className={cx("button", `btn--${kind}`, `btn--${size}`)}
  >
    {iconLeft}
    {children}
  </button>
);

export default Button;
