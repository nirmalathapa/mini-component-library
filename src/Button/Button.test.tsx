import React from "react";
import { render, screen } from "@testing-library/react";
import Button from "./Button";
import { ButtonProps } from "./Button.types";

describe("Test Component", () => {
  let props: ButtonProps;

  beforeEach(() => {
    props = {
      size: "sm",
    };
  });

  const renderComponent = () => render(<Button {...props} />);
  it("renders the component with snapshot", () => {
    const { asFragment } = render(<Button {...props}>Submit</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render Submit text correctly", () => {
    renderComponent();
    const component = screen.getByTestId("button");
    expect(component).toHaveTextContent("Submit");
  });

  it("should have small button size as class name", () => {
    props.size = "sm";
    renderComponent();
    const component = screen.getByTestId("button");
    expect(component).toHaveClass("button btn--primary btn--sm");
  });
});
