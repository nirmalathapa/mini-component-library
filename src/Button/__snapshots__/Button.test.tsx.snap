// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Component renders the component with snapshot 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
    class="button btn--primary btn--sm"
    data-testid="button"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="none"
      focusable="false"
      height="20"
      viewBox="0 0 14 14"
      width="20"
    >
      <path
        clip-rule="evenodd"
        d="M8 1.75v1a.25.25 0 0 1-.25.25h-3.5A.25.25 0 0 1 4 2.75v-1a.25.25 0 0 1 .25-.25h3.5a.25.25 0 0 1 .25.25Z"
        stroke="var(--white)"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
      <path
        d="M8 2h1.483a1 1 0 0 1 1 1v6.5a1 1 0 0 1-1 1H2.5a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1H4v0m.5 3.5h3m-3 2H6"
        stroke="var(--white)"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    Submit
  </button>
</DocumentFragment>
`;
